lib/maths/unittest/CBoostedTreeUtilsTest.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+
+#include <maths/CBoostedTree.h>
+#include <maths/CBoostedTreeFactory.h>
+#include <maths/CBoostedTreeImpl.h>
+#include <maths/CBoostedTreeLoss.h>
+#include <maths/CBoostedTreeUtils.h>
+
+#include <test/CRandomNumbers.h>
+
+#include "BoostedTreeTestData.h"
+#include "maths/COrderings.h"
+
+#include <boost/test/unit_test.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <map>
+#include <utility>
+#include <vector>
+
+BOOST_AUTO_TEST_SUITE(CBoostedTreeUtilsTest)
+
+using namespace ml;
+using namespace maths::boosted_tree_detail;
+using TDoubleVec = std::vector<double>;
+using TDoubleVecVec = std::vector<TDoubleVec>;
+
+BOOST_AUTO_TEST_CASE(testRetrainTreeSelectionProbabilities) {
+
+    std::size_t rows{100};
+    std::size_t cols{4};
+    TDoubleVec m{10.0, 2.0, 5.0};
+    TDoubleVec s{1.0, -0.9, 0.8};
+    double noiseVariance{1.0};
+
+    TTargetFunc target{[=](const core::CDataFrame::TRowRef& row) {
+        double result{0.0};
+        for (std::size_t i = 0; i < cols - 1; ++i) {
+            result += m[i] + s[i] * row[i];
+        }
+        return result;
+    }};
+
+    test::CRandomNumbers rng;
+
+    auto frame = setupRegressionProblem(rng, target, noiseVariance, rows, cols);
+
+    auto regression = maths::CBoostedTreeFactory::constructFromParameters(
+                          1, std::make_unique<maths::boosted_tree::CMse>())
+                          .buildFor(*frame, cols - 1);
+    regression->train();
+
+    const auto& impl = regression->impl();
+
+    // Edge case.
+    {
+        core::CPackedBitVector allRowsMask{rows, true};
+        std::vector<std::vector<maths::CBoostedTreeNode>> emptyForest;
+        auto probabilities = retrainTreeSelectionProbabilities(
+            1, *frame, impl.extraColumns(), cols - 1, impl.encoder(),
+            allRowsMask, impl.loss(), emptyForest);
+        BOOST_TEST_REQUIRE(probabilities.empty());
+    }
+
+    // Test some invariants of the probabilities.
```

Comment:
Can you please elaborate more, which invariants are tested?

Reply:
See [ea93475](https://github.com/elastic/ml-cpp/pull/1820/commits/ea934756759014d742ec625bde938a3d8cb24eb0).